src/Compilers/Core/Portable/InternalUtilities/StackGuard.cs

```diff
@@ -29,5 +30,27 @@ public static void EnsureSufficientExecutionStack(int recursionDepth)
                 RuntimeHelpers.EnsureSufficientExecutionStack();
             }
         }
+
+        internal static bool TryEnsureSufficientExecutionStack(int recursionDepth, bool throwOnFailure)
+        {
+            try
+            {
+                EnsureSufficientExecutionStack(recursionDepth);
+                return true;
+            }
+            catch (InsufficientExecutionStackException) when (!throwOnFailure)
+            {
+                return false;
+            }
+        }
+
+        internal static TResult ExecuteOnNewExecutionStack<TResult>(Func<TResult> execute)
```

Comment:
Do we want to give this a `TParam` to allow us to avoid capturing lambdas, or do you think it's not worth it here?

Reply:
`Task.Run()` takes a `Func<TResult>` so any capturing is unavoidable, and the delegate passed to `ExecuteOnNewExecutionStack()` should only be created on code paths that require a `Task`.

---
In reply to: [358472339](https://github.com/dotnet/roslyn/pull/40367#discussion_r358472339) [](ancestors = 358472339)

Reply:
> Task.Run() takes a Func<TResult> so any capturing is unavoidable

`Task.Run` is the equivalent of `Task.Factory.StartNew(..., CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default)`, and there's another overload that would let you pass in `object state` as well (along with all the same arguments).

However, you're only going to get here when you're so deep on the call stack that you have to hop to a new stack, and pay all the relevant costs associated with that.  I highly doubt the extra closure/delegate allocation is going to matter at all.

Reply:
> I highly doubt the extra closure/delegate allocation is going to matter at all

Yeah, that was my feeling as well, but we do have a general principle in the compiler of trying to avoid all capturing lambdas. So I just wanted to point it out.

---
In reply to: [358913409](https://github.com/dotnet/roslyn/pull/40367#discussion_r358913409) [](ancestors = 358913409)